### ISSUE TITLE
2018-09-02 deeplab + edge summary

### ISSUE THREAD

**Issue by yzbx** [ISCAS007/torchseg#16]
# todo
- [ ] graph summary
- [ ] dynamic edge width
- [ ] dynamic edge loss weight
- [ ] ignore label for edge

### ignore label for edge
![image](https://user-images.githubusercontent.com/5005182/44962583-1a1d2a00-af55-11e8-82db-514bcecbf96e.png)
![image](https://user-images.githubusercontent.com/5005182/44962594-3faa3380-af55-11e8-814b-3ba9da00c825.png)
![image](https://user-images.githubusercontent.com/5005182/44962597-4d5fb900-af55-11e8-930f-f982e8f11e9f.png)
![image](https://user-images.githubusercontent.com/5005182/44962606-636d7980-af55-11e8-940e-eda1b2931583.png)
![image](https://user-images.githubusercontent.com/5005182/44962605-5d779880-af55-11e8-93d2-98213019e8c5.png)

**Comment by yzbx** [comment]
### ignore label for edge (orange line for psp_edge, green line for pspnet)
![image](https://user-images.githubusercontent.com/5005182/44962626-a4658e00-af55-11e8-8ba6-c6bd71b892f6.png)

**Comment by yzbx** [comment]
# slim reference
- [eval_image_classifier.py](https://github.com/tensorflow/models/slim/eval_image_classifier.py)
- [train_image_classifier.py](https://github.com/tensorflow/models/slim/train_image_classifier.py)

# slim
- dataset_factory
- preprocessing_factory
- nets_factory: can export interface for placeholder
```
with tf.Graph().as_default() as graph:
    dataset = dataset_factory.get_dataset(FLAGS.dataset_name, 'train',
                                          FLAGS.dataset_dir)
    network_fn = nets_factory.get_network_fn(
        FLAGS.model_name,
        num_classes=(dataset.num_classes - FLAGS.labels_offset),
        is_training=FLAGS.is_training)
    image_size = FLAGS.image_size or network_fn.default_image_size
    placeholder = tf.placeholder(name='input', dtype=tf.float32,
                                 shape=[FLAGS.batch_size, image_size,
                                        image_size, 3])
    network_fn(placeholder)
    graph_def = graph.as_graph_def()
    with gfile.GFile(FLAGS.output_file, 'wb') as f:
      f.write(graph_def.SerializeToString())
```

# data flow
1. dataset_factory->[height,width,channel]
2. slim.dataset_data_provider.DatasetDataProvider->[height,width,channel]
3. preprocessing_factory->[height,width,channel]
4. tf.train.batch->[batch,height,width,channel]
5. nets_factory->[batch,height,width,channel]

# slim metrics
- tf.summary.scalar(summary_name, value, collections=[])
  - tf.add_to_collection(tf.GraphKeys.SUMMARIES,op)
```
# Define the metrics:
names_to_values, names_to_updates = slim.metrics.aggregate_metric_map({
    'Accuracy': slim.metrics.streaming_accuracy(predictions, labels),
    'Recall_5': slim.metrics.streaming_recall_at_k(
        logits, labels, 5),
})

# Print the summaries to screen.
for name, value in names_to_values.items():
  summary_name = 'eval/%s' % name
  op = tf.summary.scalar(summary_name, value, collections=[])
  op = tf.Print(op, [value], summary_name)
  tf.add_to_collection(tf.GraphKeys.SUMMARIES, op)
```

# slim init weight from checkpoint path
```
  variables_to_restore = []
  for var in slim.get_model_variables():
    for exclusion in exclusions:
      if var.op.name.startswith(exclusion):
        break
    else:
      variables_to_restore.append(var)

  init_fn=slim.assign_from_checkpoint_fn(
      checkpoint_path,
      variables_to_restore,
      ignore_missing_vars=FLAGS.ignore_missing_vars)
```
